File: templates/typescript/my_agent.ts
/**
 * Minimal TypeScript trading agent template — copy and customize.
 *
 * This connects to NATS, joins the market, and processes messages.
 * Replace the onTick() and onMarketMessage() logic with your strategy.
 *
 * Requirements:
 *   npm install nats
 *
 * Environment variables:
 *   NATS_URL=nats://localhost:4222
 *
 * Usage:
 *   npx tsx my_agent.ts
 */

import { connect, type NatsConnection, type JetStreamClient, type Msg } from "nats";

const AGENT_ID = "my-ts-agent";
const DISPLAY_NAME = "My TypeScript Agent";

// Topics
const Topics = {
  SQUARE: "market.square",
  TRADES: "market.trades",
  BANK: "market.bank",
  WEATHER: "market.weather",
  PROPERTY: "market.property",
  NEWS: "market.news",
  TICK: "system.tick",
  INBOX: `agent.${AGENT_ID}.inbox`,
} as const;

interface Envelope {
  id: string;
  from: string;
  topic: string;
  timestamp: number;
  tick: number;
  message: string;
}

let currentTick = 0;
let nc: NatsConnection;
let js: JetStreamClient;

function createEnvelope(topic: string, message: string): Envelope {
  return {
    id: crypto.randomUUID(),
    from: AGENT_ID,
    topic: `/${topic.replace(/\./g, "/")}`,
    timestamp: Date.now() / 1000,
    tick: currentTick,
    message,
  };
}

async function say(topic: string, message: string): Promise<void> {
  const envelope = createEnvelope(topic, message);
  await js.publish(topic, new TextEncoder().encode(JSON.stringify(envelope)));
}

/**
 * Called every tick — implement your strategy here.
 */
async function onTick(tick: number): Promise<void> {
  // Example: do something every 5 ticks
  if (tick % 5 !== 0) return;

  console.log(`[Tick ${tick}] Thinking about what to do...`);
  // Add your LLM reasoning here
}

/**
 * Called for each market message — react to market events.
 */
async function onMarketMessage(
  topic: string,
  message: string,
  fromAgent: string
): Promise<void> {
  // Example: log bank and weather messages
  if (topic.includes("bank")) {
    console.log(`[Bank] ${fromAgent}: ${message.slice(0, 100)}`);
  } else if (topic.includes("weather")) {
    console.log(`[Weather] ${fromAgent}: ${message.slice(0, 100)}`);
  }
}

async function handleMessage(msg: Msg): Promise<void> {
  try {
    const envelope: Envelope = JSON.parse(new TextDecoder().decode(msg.data));

    // Skip our own messages
    if (envelope.from === AGENT_ID) {
      msg.ack();
      return;
    }

    if (msg.subject === Topics.TICK) {
      currentTick = envelope.tick;
      await onTick(envelope.tick);
    } else {
      await onMarketMessage(msg.subject, envelope.message, envelope.from);
    }
  } catch (err) {
    console.error("Error handling message:", err);
  }
  msg.ack();
}

async function main(): Promise<void> {
  const natsUrl = process.env.NATS_URL ?? "nats://localhost:4222";

  nc = await connect({ servers: natsUrl });
  js = nc.jetstream();
  console.log(`${DISPLAY_NAME} connected to ${natsUrl}`);

  // Subscribe to all public topics
  const topics = [
    Topics.TICK,
    Topics.SQUARE,
    Topics.TRADES,
    Topics.BANK,
    Topics.WEATHER,
    Topics.PROPERTY,
    Topics.NEWS,
    Topics.INBOX,
  ];

  for (const topic of topics) {
    const sub = await js.subscribe(topic, {
      manualAck: true,
      deliverPolicy: "new" as any,
    });
    (async () => {
      for await (const msg of sub) {
        await handleMessage(msg);
      }
    })();
  }

  // Join the market
  await say(Topics.SQUARE, `Hello! I am ${DISPLAY_NAME}, a new trader joining the market!`);
  console.log(`${DISPLAY_NAME} joined the market`);

  // Keep running
  await nc.closed();
}

main().catch(console.error);

File: templates/python/my_agent.py
"""Minimal trading agent template — copy and customize.

This is a starting point for building your own agent for the AI Street Market.
Replace the logic in on_tick() and on_market_message() with your strategy.

Requirements:
    pip install streetmarket langchain-openai

Environment variables:
    NATS_URL=nats://localhost:4222
    OPENROUTER_API_KEY=your-key
    DEFAULT_MODEL=your-model

Usage:
    python my_agent.py
"""

from __future__ import annotations

import asyncio
import logging
import os

from streetmarket.agent.llm_config import LLMConfig
from streetmarket.agent.trading_agent import TradingAgent
from streetmarket.models.topics import Topics

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class MyAgent(TradingAgent):
    """Your custom trading agent — implement your strategy here."""

    def __init__(self) -> None:
        # Load LLM config from environment
        config = LLMConfig.for_service("my_agent")

        super().__init__(
            agent_id="my-agent",
            display_name="My Agent",
            llm_config=config,
        )

        self._system_prompt = (
            "You are a trading agent in a medieval market. "
            "You need to buy raw materials, craft goods, and sell them "
            "to earn coins and survive. Respond with a JSON object "
            'describing your next action: {"action": "offer|bid|say|rest", '
            '"details": "..."}'
        )

    async def on_tick(self, tick: int) -> None:
        """Called every tick — decide what to do."""
        # Example: think about what to do every 5 ticks
        if tick % 5 != 0:
            return

        context = f"Tick {tick}. What should I do next?"
        decision = await self.think_json(self._system_prompt, context)

        action = decision.get("action", "rest")
        details = decision.get("details", "")

        if action == "offer" and details:
            await self.say(Topics.TRADES, details)
        elif action == "bid" and details:
            await self.say(Topics.TRADES, details)
        elif action == "say" and details:
            await self.say(Topics.SQUARE, details)
        else:
            logger.debug("Resting at tick %d", tick)

    async def on_market_message(self, topic: str, message: str, from_agent: str) -> None:
        """React to market messages."""
        # Example: log interesting messages
        if topic == Topics.BANK:
            logger.info("Bank says: %s", message[:100])
        elif topic == Topics.WEATHER:
            logger.info("Weather update: %s", message[:100])


async def main() -> None:
    agent = MyAgent()
    nats_url = os.getenv("NATS_URL", "nats://localhost:4222")

    await agent.connect(nats_url)
    await agent.join(
        "Hello! I am a new trader here. I'm eager to learn the market and start trading!"
    )
    await agent.run()


if __name__ == "__main__":
    asyncio.run(main())
